operator/builtin/input/file/file.go

```diff
@@ -111,55 +118,67 @@ func (f *InputOperator) startPoller(ctx context.Context) {
 	}()
 }
 
+const RotatedFileTrackingLimit = 10
+
 // poll checks all the watched paths for new entries
 func (f *InputOperator) poll(ctx context.Context) {
-	var matches []string
-	if len(f.queuedMatches) > f.MaxConcurrentFiles {
-		matches, f.queuedMatches = f.queuedMatches[:f.MaxConcurrentFiles], f.queuedMatches[f.MaxConcurrentFiles:]
-	} else {
-		if len(f.queuedMatches) > 0 {
-			matches, f.queuedMatches = f.queuedMatches, make([]string, 0)
-		} else {
-			// Increment the generation on all known readers
-			// This is done here because the next generation is about to start
-			for i := 0; i < len(f.knownFiles); i++ {
-				f.knownFiles[i].generation++
-			}
-
-			// Get the list of paths on disk
-			matches = getMatches(f.Include, f.Exclude)
-			if f.firstCheck && len(matches) == 0 {
-				f.Warnw("no files match the configured include patterns", "include", f.Include)
-			} else if len(matches) > f.MaxConcurrentFiles {
-				matches, f.queuedMatches = matches[:f.MaxConcurrentFiles], matches[f.MaxConcurrentFiles:]
-			}
-		}
+	matches := getMatches(f.Include, f.Exclude)
+	if f.firstCheck && len(matches) == 0 {
+		f.Warnw("no files match the configured include patterns", "include", f.Include)
 	}
-
-	// Open the files first to minimize the time between listing and opening
-	files := make([]*os.File, 0, len(matches))
 	for _, path := range matches {
-		if _, ok := f.SeenPaths[path]; !ok {
-			if f.startAtBeginning {
-				f.Infow("Started watching file", "path", path)
-			} else {
-				f.Infow("Started watching file from end. To read preexisting logs, configure the argument 'start_at' to 'beginning'", "path", path)
-			}
-			f.SeenPaths[path] = struct{}{}
-		}
-		file, err := os.Open(path)
+		reader, err := f.isNewFileToTail(path)
 		if err != nil {
-			f.Errorw("Failed to open file", zap.Error(err))
 			continue
 		}
-		files = append(files, file)
-	}
+		if reader == nil {
+			continue
+		}
 
-	readers := f.makeReaders(files)
+		if _, ok := f.tailingReaders[path]; !ok { // If path doesn't exist => new file to tail
+			if f.firstCheck {
+				if f.startAtBeginning {
+					f.Infow("Started watching file", "path", path)
+				} else {
+					f.Infow("Started watching file from end. To read preexisting logs, configure the argument 'start_at' to 'beginning'", "path", path)
+				}
+				if err := reader.InitializeOffset(f.startAtBeginning); err != nil {
+					f.Errorw("Failed to initialize offset for "+path, zap.Error(err))
+				}
+			} else {
+				f.Infow("Started watching file", "path", path)
+			}
+			f.tailingReaders[path] = []*Reader{reader}
+		} else { // If path exists
+			f.Infow("Log rotation detected. Started watching file", "path", path)
```

Comment:
Does this necessarily mean the file has rotated? Wouldn't this happen any time we have seen the same file path before?

Reply:
in every `poll` cycle, we will be checking the same file we have been tailing. only if the fingerprint is different and the path is what we have been tailing, then it means log rotation. 
Did i address your concern well here?